src/core/shared_memory_manager.cc

```diff
@@ -0,0 +1,244 @@
+    for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
+         ++itr) {
+      if (itr->second->shm_key_ == shm_key)
+        shm_fd = itr->second->shm_fd_;
+      else {
+        RETURN_IF_ERROR(OpenSharedMemoryRegion(shm_key, &shm_fd));
+      }
+    }
+
+    RETURN_IF_ERROR(
+        GetSharedMemoryAddress(shm_fd, offset, byte_size, &tmp_addr));
+    shared_memory_map_.insert(std::make_pair(
+        name, std::unique_ptr<SharedMemoryInfo>(new SharedMemoryInfo(
+                  name, shm_key, offset, byte_size, shm_fd, tmp_addr))));
+  }
+  catch (std::exception& ex) {
+    return Status(
+        RequestStatusCode::INVALID_ARG,
+        "Failed to register shared memory region '" + name + "'.");
+  }
+
+  return Status::Success;
+}
+
+Status
+SharedMemoryManager::UnregisterSharedMemory(const std::string& name)
+{
+  LOG_VERBOSE(1) << "Unregister() shared memory region: '" << name << "'";
+
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(register_mu_);
+
+  auto it = shared_memory_map_.find(name);
+  if (it != shared_memory_map_.end()) {
+    RETURN_IF_ERROR(MmapCleanup(it->second->shm_addr_, it->second->byte_size_));
+
+    // if no other region with same shm_key then close
+    bool last_one = true;
+    for (auto itr = shared_memory_map_.begin(); itr != shared_memory_map_.end();
+         ++itr) {
+      if (itr->second->shm_key_ == it->second->shm_key_) {
+        last_one = false;
+        break;
+      }
+    }
+    if (last_one) {
+      RETURN_IF_ERROR(CloseSharedMemoryRegion(it->second->shm_fd_));
+    }
+
+    // remove region info from shared_memory_map_
+    shared_memory_map_.erase(it);
+  }
+
+  return Status::Success;
+}
+
+SharedMemoryManager::SharedMemoryManager(
+    const std::shared_ptr<ServerStatusManager>& status_manager)
+    : status_manager_(status_manager)
+{
+}
+
+SharedMemoryManager::~SharedMemoryManager()
+{
+  UnregisterAllSharedMemory();
+}
+
+Status
+SharedMemoryManager::Create(
+    const std::shared_ptr<ServerStatusManager>& status_manager,
+    std::unique_ptr<SharedMemoryManager>* shared_memory_manager)
+{
+  // Not setting the smart pointer directly to simplify clean up
+  std::unique_ptr<SharedMemoryManager> tmp_manager(
+      new SharedMemoryManager(status_manager));
+  *shared_memory_manager = std::move(tmp_manager);
+
+  return Status::Success;
+}
+
+Status
+SharedMemoryManager::UnregisterAllSharedMemory()
+{
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(register_mu_);
+
+  std::string error_message =
+      "Failed to unregister the following shared memory regions: ";
+  std::vector<std::string> unregister_fails;
+  for (const auto& shm_info : shared_memory_map_) {
+    Status unregister_status = UnregisterSharedMemory(shm_info.first);
+    if (!unregister_status.IsOk()) {
+      unregister_fails.push_back(shm_info.first);
```

Comment:
Since we aren't returning the actual 'unregister_status' error message you should log it with LOG_ERROR